File: packages/python/tests/integration/binding/scan_failures_test.py
"""Binding-tier tests for the scan's record of files it skipped (#715).

Since #714 a file whose ``on_file`` hook raises, or whose row the table
rejects, is skipped rather than failing the scan. The CLI reports those skips
on stderr and exits 23; a Python caller had no equivalent, so an incomplete
index was indistinguishable from a complete one -- the regression this closes.
"""

import os

import pytest

from dirsql import DirSQL, Table


def _db(tmp_dir, on_file, ddl="CREATE TABLE items (name TEXT)", **kw):
    return DirSQL(
        tmp_dir,
        tables=[Table(ddl=ddl, glob="*.json", on_file=on_file, **kw)],
    )


def _write(tmp_dir, *names):
    for name in names:
        with open(os.path.join(tmp_dir, name), "w") as f:
            f.write("{}")


def describe_scan_failures():
    @pytest.mark.asyncio
    async def it_is_empty_after_a_clean_scan(tmp_dir):
        _write(tmp_dir, "a.json")
        db = _db(tmp_dir, lambda path: [{"name": "ok"}])
        await db.ready()
        assert await db.scan_failures() == []

    @pytest.mark.asyncio
    async def it_names_each_skipped_file(tmp_dir):
        _write(tmp_dir, "good.json", "bad.json")

        def on_file(path):
            if path.endswith("bad.json"):
                raise ValueError("boom")
            return [{"name": "ok"}]

        db = _db(tmp_dir, on_file)
        await db.ready()

        failures = await db.scan_failures()
        assert len(failures) == 1, failures
        assert failures[0].path.endswith("bad.json"), failures[0].path
        # The row that did land is untouched: this reports, it does not gate.
        assert await db.query("SELECT name FROM items") == [{"name": "ok"}]

    @pytest.mark.asyncio
    async def it_carries_the_hooks_own_message(tmp_dir):
        _write(tmp_dir, "a.json")
        db = _db(tmp_dir, lambda path: (_ for _ in ()).throw(ValueError("boom-xyzzy")))
        await db.ready()

        (failure,) = await db.scan_failures()
        assert "boom-xyzzy" in failure.message, failure.message

    @pytest.mark.asyncio
    async def it_reports_a_row_the_table_rejected(tmp_dir):
        # Not just hook exceptions: a strict-mode violation is the same kind of
        # per-file failure, and the message still names the offending column.
        _write(tmp_dir, "a.json")
        db = _db(tmp_dir, lambda path: [{"nope": 1}], strict=True)
        await db.ready()

        (failure,) = await db.scan_failures()
        assert failure.path.endswith("a.json"), failure.path
        assert "nope" in failure.message, failure.message

    @pytest.mark.asyncio
    async def it_reports_every_skipped_file_not_only_the_first(tmp_dir):
        _write(tmp_dir, "a.json", "b.json", "c.json")
        db = _db(tmp_dir, lambda path: (_ for _ in ()).throw(ValueError("boom")))
        await db.ready()

        reported = {os.path.basename(f.path) for f in await db.scan_failures()}
        assert reported == {"a.json", "b.json", "c.json"}, reported

File: packages/ts/tests/integration/binding/scan-failures.test.ts
// Binding-tier tests for the scan's record of files it skipped (#715).
//
// Since #714 a file whose `onFile` hook throws, or whose row the table
// rejects, is skipped rather than failing the scan. The CLI reports those
// skips on stderr and exits 23; a TypeScript caller had no equivalent, so an
// incomplete index was indistinguishable from a complete one -- the
// regression this closes.
import { writeFileSync } from "node:fs";
import { mkdtemp, rm } from "node:fs/promises";
import { tmpdir } from "node:os";
import { basename, join } from "node:path";
import { DirSQL } from "dirsql";
import { afterEach, beforeEach, describe, expect, it } from "vitest";

describe("DirSQL scanFailures", () => {
  let dir: string;

  beforeEach(async () => {
    dir = await mkdtemp(join(tmpdir(), "dirsql-scan-failures-"));
  });

  afterEach(async () => {
    await rm(dir, { recursive: true, force: true });
  });

  const write = (...names: string[]) => {
    for (const name of names) {
      writeFileSync(join(dir, name), "{}");
    }
  };

  const mkdb = (
    onFile: (filePath: string) => unknown[],
    extra: Record<string, unknown> = {},
  ) =>
    new DirSQL({
      root: dir,
      tables: [
        {
          name: "items",
          ddl: "CREATE TABLE items (name TEXT)",
          glob: "*.json",
          onFile,
          ...extra,
        },
      ],
    });

  it("is empty after a clean scan", async () => {
    write("a.json");
    const db = mkdb(() => [{ name: "ok" }]);
    await db.ready;
    expect(await db.scanFailures()).toEqual([]);
  });

  it("names each skipped file", async () => {
    write("good.json", "bad.json");
    const db = mkdb((filePath) => {
      if (filePath.endsWith("bad.json")) {
        throw new Error("boom");
      }
      return [{ name: "ok" }];
    });
    await db.ready;

    const failures = await db.scanFailures();
    expect(failures).toHaveLength(1);
    expect(basename(failures[0].path)).toBe("bad.json");
    // The row that did land is untouched: this reports, it does not gate.
    expect(await db.query("SELECT name FROM items")).toEqual([{ name: "ok" }]);
  });

  it("carries the hook's own message", async () => {
    write("a.json");
    const db = mkdb(() => {
      throw new Error("boom-xyzzy");
    });
    await db.ready;

    const [failure] = await db.scanFailures();
    expect(failure.message).toContain("boom-xyzzy");
  });

  it("reports a row the table rejected", async () => {
    // Not just thrown hooks: a strict-mode violation is the same kind of
    // per-file failure, and the message still names the offending column.
    write("a.json");
    const db = mkdb(() => [{ nope: 1 }], { strict: true });
    await db.ready;

    const [failure] = await db.scanFailures();
    expect(basename(failure.path)).toBe("a.json");
    expect(failure.message).toContain("nope");
  });

  it("reports every skipped file, not only the first", async () => {
    write("a.json", "b.json", "c.json");
    const db = mkdb(() => {
      throw new Error("boom");
    });
    await db.ready;

    const names = (await db.scanFailures()).map((f) => basename(f.path)).sort();
    expect(names).toEqual(["a.json", "b.json", "c.json"]);
  });
});
